# Sentiment Analysis On Twitter

### By - Lasya Pathuri

The sentiment analysis was performed on a "word" level. If the number of positive words in all of the tweets were more than the number of negative words, then I've concluded that the sentiment is positive. I'm calculating overall sentiment as difference of number of positive words and negative words.

### Task 1: First task is to collect tweets with word "Trump" to perform sentiment analysis

In [256]:
# Print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### Load the files with positive, negative, stop words

In [257]:
# --------To read positive words from a text file and put it in a list--------
positive = [word.strip() for word in open('positive.txt','r') ]
print(positive)

['a+', 'abound', 'abounds', 'abundance', 'abundant', 'accessable', 'accessible', 'acclaim', 'acclaimed', 'acclamation', 'accolade', 'accolades', 'accommodative', 'accomodative', 'accomplish', 'accomplished', 'accomplishment', 'accomplishments', 'accurate', 'accurately', 'achievable', 'achievement', 'achievements', 'achievible', 'acumen', 'adaptable', 'adaptive', 'adequate', 'adjustable', 'admirable', 'admirably', 'admiration', 'admire', 'admirer', 'admiring', 'admiringly', 'adorable', 'adore', 'adored', 'adorer', 'adoring', 'adoringly', 'adroit', 'adroitly', 'adulate', 'adulation', 'adulatory', 'advanced', 'advantage', 'advantageous', 'advantageously', 'advantages', 'adventuresome', 'adventurous', 'advocate', 'advocated', 'advocates', 'affability', 'affable', 'affably', 'affectation', 'affection', 'affectionate', 'affinity', 'affirm', 'affirmation', 'affirmative', 'affluence', 'affluent', 'afford', 'affordable', 'affordably', 'afordable', 'agile', 'agilely', 'agility', 'agreeable', 'ag

In [258]:
# --------To read negative words from a text file and put it in a list--------
negative = [word.strip() for word in open('negative-words.txt','r')]
print(negative)

['2-faced', '2-faces', 'abnormal', 'abolish', 'abominable', 'abominably', 'abominate', 'abomination', 'abort', 'aborted', 'aborts', 'abrade', 'abrasive', 'abrupt', 'abruptly', 'abscond', 'absence', 'absent-minded', 'absentee', 'absurd', 'absurdity', 'absurdly', 'absurdness', 'abuse', 'abused', 'abuses', 'abusive', 'abysmal', 'abysmally', 'abyss', 'accidental', 'accost', 'accursed', 'accusation', 'accusations', 'accuse', 'accuses', 'accusing', 'accusingly', 'acerbate', 'acerbic', 'acerbically', 'ache', 'ached', 'aches', 'achey', 'aching', 'acrid', 'acridly', 'acridness', 'acrimonious', 'acrimoniously', 'acrimony', 'adamant', 'adamantly', 'addict', 'addicted', 'addicting', 'addicts', 'admonish', 'admonisher', 'admonishingly', 'admonishment', 'admonition', 'adulterate', 'adulterated', 'adulteration', 'adulterier', 'adversarial', 'adversary', 'adverse', 'adversity', 'afflict', 'affliction', 'afflictive', 'affront', 'afraid', 'aggravate', 'aggravating', 'aggravation', 'aggression', 'aggress

In [259]:
# --------To read stop words from a text file and put it in a list--------
stop = [word.strip() for word in open('stopwords.txt','r') ]
print(stop)

['about', 'above', 'across', 'after', 'again', 'against', 'all', 'almost', 'alone', 'along', 'already', 'also', 'although', 'always', 'among', 'an', 'and', 'another', 'any', 'anybody', 'anyone', 'anything', 'anywhere', 'are', 'area', 'areas', 'around', 'as', 'ask', 'asked', 'asking', 'asks', 'at', 'away', 'back', 'backed', 'backing', 'backs', 'be', 'became', 'because', 'become', 'becomes', 'been', 'before', 'began', 'behind', 'being', 'beings', 'best', 'better', 'between', 'big', 'both', 'but', 'by', 'came', 'can', 'cannot', 'case', 'cases', 'certain', 'certainly', 'clear', 'clearly', 'come', 'could', 'did', 'differ', 'different', 'differently', 'do', 'does', 'done', 'down', 'down', 'downed', 'downing', 'downs', 'during', 'each', 'early', 'either', 'end', 'ended', 'ending', 'ends', 'enough', 'even', 'evenly', 'ever', 'every', 'everybody', 'everyone', 'everything', 'everywhere', 'face', 'faces', 'fact', 'facts', 'far', 'felt', 'few', 'find', 'finds', 'first', 'for', 'four', 'from', 'ful

### Dictionary of positive,negative words

Creating a dictionary of positive, negative words and assigning value 1 to positive words, -1 to negative words

In [260]:
# --------To make a single dictionary of positive and negative words.
d = {}
# Each positive word is a key in the dictionary and its value is assigned to be +1 using above for loop
for word in positive:
    d[word] = 1
# Each negative word is a key in the dictionary and its value is assigned to be -1 using above for loop
for word in negative:
    d[word]=-1
# Prints dictionary of words as keys and values as +1 for positive word and -1 for negative word
print(d)

{'a+': 1, 'abound': 1, 'abounds': 1, 'abundance': 1, 'abundant': 1, 'accessable': 1, 'accessible': 1, 'acclaim': 1, 'acclaimed': 1, 'acclamation': 1, 'accolade': 1, 'accolades': 1, 'accommodative': 1, 'accomodative': 1, 'accomplish': 1, 'accomplished': 1, 'accomplishment': 1, 'accomplishments': 1, 'accurate': 1, 'accurately': 1, 'achievable': 1, 'achievement': 1, 'achievements': 1, 'achievible': 1, 'acumen': 1, 'adaptable': 1, 'adaptive': 1, 'adequate': 1, 'adjustable': 1, 'admirable': 1, 'admirably': 1, 'admiration': 1, 'admire': 1, 'admirer': 1, 'admiring': 1, 'admiringly': 1, 'adorable': 1, 'adore': 1, 'adored': 1, 'adorer': 1, 'adoring': 1, 'adoringly': 1, 'adroit': 1, 'adroitly': 1, 'adulate': 1, 'adulation': 1, 'adulatory': 1, 'advanced': 1, 'advantage': 1, 'advantageous': 1, 'advantageously': 1, 'advantages': 1, 'adventuresome': 1, 'adventurous': 1, 'advocate': 1, 'advocated': 1, 'advocates': 1, 'affability': 1, 'affable': 1, 'affably': 1, 'affectation': 1, 'affection': 1, 'affe

### Extracting the text file containing tweets 

In [261]:
with open('Trump.txt', 'r') as my_file:
    data = my_file.read()
print(data)

b'RT @RyanGirdusky: The rise of right-wing populism from Orban to Le Pen to Trump to Salvini... my new article\nhttps://t.co/3tnStoIBXJ\n'b'RT @HoarseWisperer: Michael Avenatti stole a parapalegic\xe2\x80\x99s settlement money and spent it on race cars.\n\nThe guy is Trump without the inheri\xe2\x80\xa6\n'b'RT @TheRickWilson: I love the overlap of the Trump-right media Assange fanboy demo with the Bernie Bro Assange fanboy demo.\n'b"RT @ggreenwald: This is Assange's lawyer, confirming he was arrested not just for bail-jumping in the UK (a minor offense) but also on a US\xe2\x80\xa6\n"b'RT @fras99: If Trump loses the 2020 election, will he leave the White House peacefully? \n\n-CNN\n'b'Tucker: We deserve to know why Trump was spied on https://t.co/xps4Xfzc7w via @YouTube\n'b'RT @Lukewearechange: One of #assange Statements from his lawyers is \xe2\x80\x9cI told you so\xe2\x80\x9d\n\nHearing also that Jeff Session was instrumental to eve\xe2\x80\xa6\n'b'RT @JohnJHarwood: this is wrong\n\n

### Saving the data as a list

In [262]:
# Prints list of tweets by splitting each sentence into set of words.
tweets = data.split(" ")
print(tweets)

["b'RT", '@RyanGirdusky:', 'The', 'rise', 'of', 'right-wing', 'populism', 'from', 'Orban', 'to', 'Le', 'Pen', 'to', 'Trump', 'to', 'Salvini...', 'my', 'new', "article\\nhttps://t.co/3tnStoIBXJ\\n'b'RT", '@HoarseWisperer:', 'Michael', 'Avenatti', 'stole', 'a', 'parapalegic\\xe2\\x80\\x99s', 'settlement', 'money', 'and', 'spent', 'it', 'on', 'race', 'cars.\\n\\nThe', 'guy', 'is', 'Trump', 'without', 'the', "inheri\\xe2\\x80\\xa6\\n'b'RT", '@TheRickWilson:', 'I', 'love', 'the', 'overlap', 'of', 'the', 'Trump-right', 'media', 'Assange', 'fanboy', 'demo', 'with', 'the', 'Bernie', 'Bro', 'Assange', 'fanboy', 'demo.\\n\'b"RT', '@ggreenwald:', 'This', 'is', "Assange's", 'lawyer,', 'confirming', 'he', 'was', 'arrested', 'not', 'just', 'for', 'bail-jumping', 'in', 'the', 'UK', '(a', 'minor', 'offense)', 'but', 'also', 'on', 'a', 'US\\xe2\\x80\\xa6\\n"b\'RT', '@fras99:', 'If', 'Trump', 'loses', 'the', '2020', 'election,', 'will', 'he', 'leave', 'the', 'White', 'House', 'peacefully?', "\\n\\n-CNN\

### Cleaning the data, Calculating the count of words and ratio:

In [263]:
import re
def cal_sentiment_analysis(tweet_data):
# Create and initialize a variable to count number of positive, negative, stop and other words
    positive_count=0
    negative_count = 0
    stop_count = 0
    other_count = 0
    Total_word_count = 0
    for word in tweet_data:
        w1 = word.lower().strip("’”\"‘ !#@,./\_][() >:';&“》©$’?•") # remove punctuations attached to each word
        text = re.sub(r'^https?:\/\/.*[\r\n]*', '', w1, flags=re.MULTILINE)   # removing URL
        Total_word_count+=1
        if(text not in stop):                                           # If word not an stop word
            if (text in d.keys()):                                      # If clean word is present in the dictionary
                if (d[text] == 1):                                      # if it is a positive word
                    positive_count += 1                                 # increase the count of positive word               
                elif (d[text] == -1):                                   # if it is a negative word
                    negative_count += 1                                 # increase the count of negative word
            else:                                                       # else
                other_count+=1                                          # increase the count of other word
        if(text in stop):
            stop_count+=1
    
    # Print number of words
    print('Total word count: ' + str(Total_word_count))              
    print('Total number of positive words: ' +str(positive_count))
    print('Total Number of negative words: ' +str(negative_count))
    print('Total number of stop words: ' +str(stop_count))
    print('Total number of other words: ' +str(other_count))
    
    # Calculating Ratios
    print('\nThe ratio of postive word compared to the whole word count is: '+str(round(positive_count/Total_word_count,3)))
    print('The ratio of negative word compared to the whole word count is: '+str(round(negative_count/Total_word_count,3)))
    print('The ratio of stop word compared to the whole word count is: '+str(round(stop_count/Total_word_count,3)))
    print('The ratio of other word compared to the whole word count is: '+str(round(other_count/Total_word_count,3)))
    
    # sum of postive and negative words
    sum_of_positive_negative = (positive_count+negative_count)
    print('The sum of all postive or negative words observed:' +str(sum_of_positive_negative))
    
    
    # Overall sentiment
    if(positive_count>negative_count):
        print('\033[1m'+'\nOverall sentiment of the tweets is a positive number:'+str(positive_count-negative_count))
        print('As the sentiment is positive, percentage of positive words: '+ str(positive_count*100/sum_of_positive_negative)+'%')
        if((positive_count*100/sum_of_positive_negative) > 50):
            print('\033[1m'+'\nWe conclude that the overall sentiment is Strongly positive')
        else:
            print('\nWe conclude that the overall sentiment is Weakly Positive')
    else:
        print('\033[1m'+'\nOverall sentiment of the tweets is a negative number:'+str(negative_count - positive_count))
        print('As the sentiment is negative, percentage of positive words: '+ str(negative_count*100/sum_of_positive_negative)+'%')
        if((negative_count*100/sum_of_positive_negative) > 50):
            print('\033[1m'+'\nWe conclude that the overall sentiment is Strongly Negative')
        else:
            print('\nWe conclude that the overall sentiment is Weakly Negative')
    


### Calling function cal_sentiment_analysis

In [264]:
cal_sentiment_analysis(tweets)

Total word count: 15614
Total number of positive words: 683
Total Number of negative words: 567
Total number of stop words: 6255
Total number of other words: 8109

The ratio of postive word compared to the whole word count is: 0.044
The ratio of negative word compared to the whole word count is: 0.036
The ratio of stop word compared to the whole word count is: 0.401
The ratio of other word compared to the whole word count is: 0.519
The sum of all postive or negative words observed:1250

Overall sentiment of the tweets is a positive number:116
As the sentiment is positive, percentage of positive words: 54.64%

We conclude that the overall sentiment is Strongly positive


### Is the general sentiment positive or negative?


The total number of positive words is: 683.

The total number of negative words is: 567.

The overall sentiment of the tweets in numeric value is: 683-567 = 116.

As the overall sentiment of the tweets is a positive number, we can say that the general sentiment is positive.

### Is the sentiment strongly or weakly positive?

Lets define the sentiment level into 4 parts:

0 % - 50: Weak

50% - 100%: Strong

The total number of words is 15614.

As the sentiment is positive, the percentage of positive words: (683/1250)*100 = 54.64%.

As 4.37% lies in the 1st part of our sentiment level, we can conclude that the overall sentiment is weak.

# Task 2: Repeating the above analysis on word of my choice

The word I have chosen to perform the analysis is "Purple"

### Extracting the text file containing tweets 

In [265]:
with open('Purple.txt', 'r') as my_file:
    data = my_file.read()
print(data)

b'RT @RockWalkLondon: Ritchie Blackmore is 74 today, born 14th April 1945 in Weston-super-Mare, Somerset, England.\nGuitarist &amp; songwriter.\nBe\xe2\x80\xa6\n'b"Let's make the sky purple! Let's go @LAGladiators! #ShieldsUP https://t.co/ke45ShzUTz\n"b'RT @BTS_AHC: To the new ARMYs:\nYou just gained millions of shoulders to cry on. Millions of ears to listen to you. Millions of arms to hug\xe2\x80\xa6\n'b'@MyOnlySanShine San / purple \xf0\x9f\x92\x9c\n'b'RT @snowberrytae: @BTS_twt MY BEAUTIFUL ANGEL I LOVE YOU SO MUCH TAEHYUNG you did so great, please always take care \xf0\x9f\x92\x9c\xf0\x9f\x92\x9c\xf0\x9f\x92\x9c purple you https://\xe2\x80\xa6\n'b'You got it right. It\xe2\x80\x99s a light purple :)\n'b'@MutantScreaming Wish the purple man would snap away my Other Country neighbs \xf0\x9f\x98\xad\xf0\x9f\x98\xb4\n'b'RT @RunOldMan: Wife went to the spa, got her nails done with purple gel, imagine my surprise when I licked her fingers and they didn\xe2\x80\x99t tast\xe2\x80\xa6\n'b'R

### Saving the data as a list

In [266]:
purple_data = data.split(" ")
print(purple_data)

["b'RT", '@RockWalkLondon:', 'Ritchie', 'Blackmore', 'is', '74', 'today,', 'born', '14th', 'April', '1945', 'in', 'Weston-super-Mare,', 'Somerset,', 'England.\\nGuitarist', '&amp;', 'songwriter.\\nBe\\xe2\\x80\\xa6\\n\'b"Let\'s', 'make', 'the', 'sky', 'purple!', "Let's", 'go', '@LAGladiators!', '#ShieldsUP', 'https://t.co/ke45ShzUTz\\n"b\'RT', '@BTS_AHC:', 'To', 'the', 'new', 'ARMYs:\\nYou', 'just', 'gained', 'millions', 'of', 'shoulders', 'to', 'cry', 'on.', 'Millions', 'of', 'ears', 'to', 'listen', 'to', 'you.', 'Millions', 'of', 'arms', 'to', "hug\\xe2\\x80\\xa6\\n'b'@MyOnlySanShine", 'San', '/', 'purple', "\\xf0\\x9f\\x92\\x9c\\n'b'RT", '@snowberrytae:', '@BTS_twt', 'MY', 'BEAUTIFUL', 'ANGEL', 'I', 'LOVE', 'YOU', 'SO', 'MUCH', 'TAEHYUNG', 'you', 'did', 'so', 'great,', 'please', 'always', 'take', 'care', '\\xf0\\x9f\\x92\\x9c\\xf0\\x9f\\x92\\x9c\\xf0\\x9f\\x92\\x9c', 'purple', 'you', "https://\\xe2\\x80\\xa6\\n'b'You", 'got', 'it', 'right.', 'It\\xe2\\x80\\x99s', 'a', 'light', 'purp

### Cleaning the data, Calculating the count of words and ratio :

In [267]:
cal_sentiment_analysis(purple_data)

Total word count: 5857
Total number of positive words: 205
Total Number of negative words: 70
Total number of stop words: 1919
Total number of other words: 3663

The ratio of postive word compared to the whole word count is: 0.035
The ratio of negative word compared to the whole word count is: 0.012
The ratio of stop word compared to the whole word count is: 0.328
The ratio of other word compared to the whole word count is: 0.625
The sum of all postive or negative words observed:275

Overall sentiment of the tweets is a positive number:135
As the sentiment is positive, percentage of positive words: 74.54545454545455%

We conclude that the overall sentiment is Strongly positive


### Is the general sentiment positive or negative?


The total number of positive words is: 205.

The total number of negative words is: 70.

The overall sentiment of the tweets in numeric value is: 683-567 = 135.

As the overall sentiment of the tweets is a positive number, we can say that the general sentiment is positive.

### Is the sentiment strongly or weakly positive?

Lets define the sentiment level into 4 parts:

0 % - 50: Weak

50% - 100%: Strong

The total number of words is 3663.

As the sentiment is positive, the percentage of positive words: (205/275)*100 = 74.54%.

As 5.59% lies in the 1st part of our sentiment level, we can conclude that the overall sentiment is Strong.